### PR TITLE
Making Tracks layers creation and update faster

### PR DESCRIPTION
**Pull request by jules-vanaret**
# References and relevant issues
I work with large tracking datasets (~10⁵ tracks, ~10⁶ vertices), and creating/setting data of a Tracks layer is rather slow, on the order of 10 seconds on my machine). I have dug into the code and found that the culprit is this function in `napari/layers/tracks/_track_utils.py`

```python
def build_tracks(self):
        """build the tracks"""

        points_id = []
        track_vertices = []
        track_connex = []

        # NOTE(arl): this takes some time when the number of tracks is large
        for idx in self.unique_track_ids:
            indices = self._vertex_indices_from_id(idx)

            # grab the correct vertices and sort by time
            vertices = self.data[indices, 1:]

            # coordinates of the text identifiers, vertices and connections
            points_id += [idx] * vertices.shape[0]
            track_vertices.append(vertices)
            track_connex.append(connex(vertices))

        self._points_id = np.array(points_id)[self._ordered_points_idx]
        self._track_vertices = np.concatenate(track_vertices, axis=0)
        self._track_connex = np.concatenate(track_connex, axis=0)
```
Note the comment above the for-loop (which I believe comes from @quantumjot ). The line `indices = self._vertex_indices_from_id(idx)`, which finds vertices indices in a lookup table (a sparse matrix), slows the loop quite a bit. I believe the loop can be circumvented entirely.

# Description
`build_tracks` is called every time a Tracks layer's data is set (so when it is created, or updated after creation). The particular point of the function is to set  `self._points_id`, `self._track_vertices`, and `self._track_connex`,  which are later fed to Vispy to display the tracks. After inspecting these quantities, I think they could be computed faster (without a for-loop).
 *  `self._points_id` is a numpy array representing the track id associated to all vertices sorted by time.
 *  `self._track_vertices` is a numpy array representing the vertices coordinates
 * `self._track_connex` is a numpy array whose length is equal to the number of vertices, and which is filled with `True` except for vertices that end a track, for which it is equal to `False`

### 1. `self._points_id`
In the current version, a list `points_id` is constructed by looping over all unique tracks indices `idx` and incrementing the list by a sub_list `[idx] * vertices.shape[0]`. The final list is then converted as a numpy array and sorted with respect to time using `np.array(points_id)[self._ordered_points_idx]`.
This can be replaced by directly defining `points_id` as `points_id = self.data[:, 0][self._ordered_points_idx].copy()`, i.e directly requesting the tracks indices from the original Tracks layer data.

### 2. self._track_vertices
Like in the previous case, a variable `vertices = self.data[indices, 1:]` is defined during the loop and appended to a list, and the latter is concatenated into `self._track_vertices`. Here, `indices` are the vertex indices, obtained by querying a lookup table. The lookup table is built in the `data.setter` of the same script as
```python
self._id2idxs = coo_matrix(
    (
        np.broadcast_to(1, self.track_ids.size),  # just dummy ones
        (self.track_ids, np.arange(self.track_ids.size)),
    )
).tocsr() 
```
Note that in the `data.setter`, the Tracks layers data `self.data` is first conveniently sorted by ID then time. This means that
`self.track_ids` (which is `self.data[:, 0].astype(np.uint32)`) is sorted. The current implementation can thus be replaced by `track_vertices = self.data[:, 1:].copy()`, i.e the vertices coordinates of the Tracks layer's original data. 

### 3. self._track_connex
Again, a list `track_connex` is created and is appended during the loop (`track_connex.append(connex(vertices))`), then concatenated to define `self._track_connex.` The function `connex` is defined as
```python
def connex(vertices: np.ndarray) -> list:
    """Connection array to build vertex edges for vispy LineVisual.

    Notes
    -----
    See
    http://api.vispy.org/en/latest/visuals.html#vispy.visuals.LineVisual

    """
    return [True] * (vertices.shape[0] - 1) + [False] 
```
Note that it uses its argument `vertices` only to compute `vertices.shape[0]`, i.e the track's duration in frames (the number of vertices in the track). `track_connex` can be defined directly by
```python
# Indices in the data array just before the track id changes
indices_new_id = np.where(np.diff(self.data[:, 0]))[0]

# Define track_connex as an array full of 'True', then set to 'False'
# at the indices just before the track id changes
track_connex = np.ones(self.data.shape[0], dtype=bool)
track_connex[indices_new_id] = False
```

# Performance changes
I have tested the performance change on my machine using the following script, which simulates a random tracking dataset with `N_vertices` vertices and `N_tracks` tracks.
```python
import numpy as np
import napari
from time import time

np.random.seed(2024)

N_vertices = 1_000_000
N_tracks = 100_000

# simulate 3D tracks (tracks are immediately sorted by track id
# and time in view_tracks/add_tracks, so we can define them randomly)
tracks = np.random.rand(N_vertices, 5) * 100
# make integer track ids
tracks[:,0] = np.random.randint(0, N_tracks, N_vertices)

tracks2 = np.random.rand(N_vertices, 5) * 100
tracks2[:,0] = np.random.randint(0, N_tracks, N_vertices)

viewer = napari.Viewer()
t0 = time()
viewer.add_tracks(tracks, name='tracks')
print(f"Creating tracks with {N_vertices} vertices, {N_tracks} tracks: {time() - t0:.4f} seconds")

t0 = time()
viewer.layers['tracks'].data = tracks2
print(f"Setting tracks with {N_vertices} vertices, {N_tracks} tracks: {time() - t0:.4f} seconds")
```
For different values of `N_vertices` and `N_tracks`, I time the following duration to define and update the Tracks layer:
|  | Old implementation  | New implementation |
| ------------- | ------------- | ------------- |
| `N_vertices = 1_000`, `N_tracks = 100`  | creation:**0.22s**, update:**0.03s**   | creation:**0.17**s, update:**0.008**s  |
| `N_vertices = 1_000_000`, `N_tracks = 10_000`  | creation:**3.49s**, update:**1.86s**   | creation:**2.81**s, update:**0.99**s  |
| `N_vertices = 1_000_000`, `N_tracks = 100_000`  | creation:**10.53s**, update:**8.94s**  | creation:**2.83s**, update:**1.03s**  |

Note that the proposed implementation has a much nicer scaling with the number of tracks and vertices, even for small values of both.
I don't know how to make this benchmark into a proper test though.

# Final Checklist
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Czaki** [comment]
could you please add your code as benchamrk to this file:
https://github.com/napari/napari/blob/main/napari/benchmarks/benchmark_tracks_layer.py

Here is part of our documentation about benchmarking https://napari.org/stable/developers/benchmarks.html

**Comment by jules-vanaret** [comment]
> could you please add your code as benchamrk to this file: https://github.com/napari/napari/blob/main/napari/benchmarks/benchmark_tracks_layer.py
> 
> Here is part of our documentation about benchmarking https://napari.org/stable/developers/benchmarks.html

I'm having an issue with _asv_. I modified the current benchmark to include an additional test for updating, and ran `asv run -E existing -b TracksSuite`, which returned
```
[ 75.00%] ··· ========= =========== =========== =========== ============
              --                            n_tracks                    
              --------- ------------------------------------------------
                 size        1           10         100         1000    
              ========= =========== =========== =========== ============
                  5       14.1±6ms      n/a         n/a         n/a     
                  50      19.1±3ms    19.8±5ms      n/a         n/a     
                 500      16.6±6ms    16.6±5ms    13.0±5ms      n/a     
                 5000    30.2±10ms    25.8±9ms    25.2±9ms    20.0±4ms  
                50000     142±40ms    120±30ms    119±30ms    109±20ms  
                500000   761±400ms   582±300ms   850±300ms   1.34±0.05s 
               5000000     failed      failed      failed      failed   
              ========= =========== =========== =========== ============

[100.00%] ··· benchmark_tracks_layer.TracksSuite.time_update_layer
[100.00%] ··· ========= ============= ============= ============ ============
              --                               n_tracks                      
              --------- -----------------------------------------------------
                 size         1             10          100          1000    
              ========= ============= ============= ============ ============
                  5       4.05±0.5ms       n/a          n/a          n/a     
                  50      4.32±0.1ms   4.23±0.03ms      n/a          n/a     
                 500     5.40±0.03ms   5.12±0.03ms   4.53±0.6ms      n/a     
                 5000     15.3±0.3ms     10.5±1ms    11.9±0.3ms   11.7±0.3ms 
                50000     96.0±0.7ms    70.9±10ms    78.4±0.7ms   79.7±0.5ms 
                500000     744±10ms      584±2ms      608±8ms      630±6ms   
               5000000      failed      5.67±0.9s    7.98±0.02s     failed   
              ========= ============= ============= ============ ============
```
Unfortunately, I cannot compare to main as `asv continuous main napari_tracks_setter_optim -b TracksSuite` returns
```
· No executable found for python 3.11
· No environments selected
```
Also I don't know if it's normal, but my `.asv` folder contains only a `results` folder (so no `.asv/env` folder path exists, as specified in `asv.conf.json`).

File: napari/layers/tracks/_track_utils.py
Comment (brisvag) [inline napari/layers/tracks/_track_utils.py]: is `copy()` actually necessary here? As far as I can tell, these arrays are never modified inplace up until they're given to vispy:

- `points_id` is only indexed via `get_value`
- `track_vertices` is returned via `_view_data`

Sometimes vispy does change arrays, so I'm not 100% sure it's fine to remove these, but since we're here maybe we can test a bit!

Comment (jules-vanaret) [reply napari/layers/tracks/_track_utils.py]: I tend to put `.copy()` instinctively now (had _many_ bad surprises in the past). Any idea how to make tests to check whether or not they are modified later down the visualization line ?

Comment (brisvag) [reply napari/layers/tracks/_track_utils.py]: I took a look into vispy and I'm 90% sure this `copy` is unneeded. I'm not sure how to test thoroughly... some things that cause issues in the past are: passing dtypes that need conversion (float64) with data outside of the range of float32; changing data inplace on the napari side (layer.data[something] = something), then triggering a `layer.refresh()`... and seeing if things go wrong ^^'

**Comment by Czaki** [comment]
I have triggered benchmark on CI. Let's wait on its result.

**Comment by jules-vanaret** [comment]
> I have triggered benchmark on CI. Let's wait on its result.

Thank you! The benchmarks seem to be positive :)
```
| Change   | Before [0747da56]    | After [20262899]    |   Ratio | Benchmark (Parameter)                                             |
|----------|----------------------|---------------------|---------|-------------------------------------------------------------------|
| -        | 3.16±0.01ms          | 1.94±0.01ms         |    0.61 | benchmark_tracks_layer.TracksSuite.time_update_layer(500, 10)     |
| -        | 2.76±0.02ms          | 1.60±0.01ms         |    0.58 | benchmark_tracks_layer.TracksSuite.time_update_layer(50, 10)      |
| -        | 24.2±0.2ms           | 13.1±0.03ms         |    0.54 | benchmark_tracks_layer.TracksSuite.time_create_layer(5000, 100)   |
| -        | 18.2±0.4ms           | 7.72±0.07ms         |    0.42 | benchmark_tracks_layer.TracksSuite.time_create_layer(500, 100)    |
| -        | 183±0.9ms            | 74.4±0.5ms          |    0.41 | benchmark_tracks_layer.TracksSuite.time_create_layer(50000, 1000) |
| -        | 15.6±0.07ms          | 4.62±0.01ms         |    0.3  | benchmark_tracks_layer.TracksSuite.time_update_layer(5000, 100)   |
| -        | 144±0.5ms            | 35.0±0.4ms          |    0.24 | benchmark_tracks_layer.TracksSuite.time_update_layer(50000, 1000) |
| -        | 12.6±0.07ms          | 1.97±0ms            |    0.16 | benchmark_tracks_layer.TracksSuite.time_update_layer(500, 100)    |
| -        | 116±0.7ms            | 13.5±0.05ms         |    0.12 | benchmark_tracks_layer.TracksSuite.time_create_layer(5000, 1000)  |
| -        | 108±0.3ms            | 4.92±0.04ms         |    0.05 | benchmark_tracks_layer.TracksSuite.time_update_layer(5000, 1000)  |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
The parameters range shown in the results above (`10<N_tracks<1000`, `50<N_tracks<50_000`) don't quite reflect the performance increase for large tracking datasets (i.e a `N_tracks > 100_000`) though.

**Comment by psobolewskiPhD** [comment]
I can confirm the bench code above is *markedly* faster on my machine locally.
With main:
```
Creating tracks with 1000000 vertices, 100000 tracks: 6.6072 seconds
Setting tracks with 1000000 vertices, 100000 tracks: 5.1887 seconds
```
with this:
```
Creating tracks with 1000000 vertices, 100000 tracks: 2.4441 seconds
Setting tracks with 1000000 vertices, 100000 tracks: 1.0332 seconds
```

**Comment by Czaki** [comment]
@jules-vanaret did you wait on any specific feedback, or require more explanation in the context of @brisvag request?

**Comment by jules-vanaret** [comment]
Sorry about the lack of news on my part, I had a busy week and will not be able to implement @brisvag's tests until next week (probably on monday).

**Comment by Czaki** [comment]
Ok. We will wait. I worring that something stuck in communication.

**Comment by brisvag** [comment]
> Sorry about the lack of news on my part, I had a busy week and will not be able to implement @brisvag's tests until next week (probably on monday).

Yeah, no rush, no worries :)

**Comment by jules-vanaret** [comment]
@brisvag I've removed the two `.copy()` and added tests to (i) check if `float64` data is handled correctly and (ii) check whether modifying a Tracks layer data inplace and calling `.refresh()` raises an error or not.
For my `float64` test, I am not sure it covers what you asked for. The test currently fails at the line `time = np.round(self._points[:, 0]).astype(np.uint)`, which does indeed cast the problematic `float64` value that I chose to the wrong value.

**Comment by brisvag** [comment]
Does the data on the canvas and in the input array still look the same at the end in both cases?

The reason why I suggested the `float64` test is because in the past we had issues where the input array was being modified inplace (somewhere deep in vispy) to fit the new dtype, and then converted, which unexpectedly changed your input array. If this doesn't happen, we're good on that front and we can remove the `copy()`.

**Comment by jules-vanaret** [comment]
> Does the data on the canvas and in the input array still look the same at the end in both cases?
> 
> The reason why I suggested the `float64` test is because in the past we had issues where the input array was being modified inplace (somewhere deep in vispy) to fit the new dtype, and then converted, which unexpectedly changed your input array. If this doesn't happen, we're good on that front and we can remove the `copy()`.

Ok I see, then my test was not appropriate!
Current test I implemented in `napari/layers/tracks/_tests/test_tracks.py` is
```python
def test_float64_tracks_data() -> None:
    """Test if float64 data is correctly handled."""
    # array with float64 values
    data = np.array([[1, 0, 0, 0, 0], [1, 10**308, 1, 1, 1]], dtype=np.float64) 
    layer = Tracks(data)

    np.testing.assert_array_equal(layer.data, data)
```
but since we want to test if Vispy modifies the array, I imagine it should rather be something like
```python
def test_float64_tracks_data(make_napari_viewer):
    """Test if float64 data is not modified inplace."""
    # array with float64 values
    data = np.array([[1, 0, 0, 0, 0], [1, 10**308, 1, 1, 1]], dtype=np.float64) 
    data_copy = data.copy()

    viewer = make_napari_viewer()
    layer = viewer.add_tracks(data)

    np.testing.assert_array_equal(data, data_copy)
```
and then I it should probably belong in `napari/_vispy/_tests/test_vispy_vectors_layer.py`. What do you think ? Is there another way to test Vispy-related behaviour without creating a Viewer instance ?

**Comment by brisvag** [comment]
I actually don't think we need to have a test in the test suite for this; I'm just saying we should manually try before merging to make sure this doesn't happen!

**Comment by jules-vanaret** [comment]
Ok, then using the following check,  I think we are good:
```python
import napari
import numpy as np

# float 64 spatial coordinate value 
tracks1 = np.array([[1, 0, 0, 0, 0], [1, 1, 10**-320, 0, 0]], dtype=np.float64) 
tracks1_copy = tracks1.copy()

# float 64 time value
tracks2 = np.array([[1, 0, 0, 0, 0], [1, 10**-320, 0, 0, 0]], dtype=np.float64) 
tracks2_copy = tracks2.copy()

# float 64 track ID value
tracks3 = np.array([[1, 0, 0, 0, 0], [10**308, 1, 0, 0, 0]], dtype=np.float64) 
tracks3_copy = tracks3.copy()

viewer = napari.Viewer()
viewer.add_tracks(tracks1)
viewer.add_tracks(tracks2)
viewer.add_tracks(tracks3)

print(np.array_equal(tracks1, tracks1_copy))
print(np.array_equal(tracks2, tracks2_copy))
print(np.array_equal(tracks3, tracks3_copy))
```
This checks returns `True` for the three `array_equal`. I switched to a small float64 value (`10**-320`) because coordinates that result in a slider (time and space coordinates in this case) are bounded between -2147483648 and 2147483648 and return an error otherwise. For the track ID one, the code raises a `RuntimeWarning: invalid value encountered in cast` because some line in `_track_utils.py` tries to convert IDs to `np.uint32`. This does not modify data inplace.

I will now remove the unnecessary test and clean `_track_utils.py` (I had left a `build_tracks_old` implementation).

**Review by brisvag — APPROVED** [review]
Awesome, thanks for doing the dirty work! I would say let's get this in then :)

**Comment by jules-vanaret** [comment]
Does the error during the latest tests point to a flaw in the PR ?

**Comment by Czaki** [comment]
No. Benchmark fail is connected with bug in benchmark configuration and upload action in version 4. I will fix it in separate PR.

**Comment by jules-vanaret** [comment]
> No. Benchmark fail is connected with bug in benchmark configuration and upload action in version 4. I will fix it in separate PR.

Ok I see. 
Well, @Czaki @brisvag @psobolewskiPhD, thank you very much for helping me through this PR ! I learnt a lot about the inner working of the Tracks layer and about the benchmarking suite. Cheers

**Comment by jules-vanaret** [comment]
Sorry to bother you again. After visually inspecting several tracking cases to make sure my implementation was not faulty I realized that I had made a big mistake. I have pushed a fix, and made a test to make sure it does not happen again. Is it still time, or should I open a new PR immediately ?

**Comment by Czaki** [comment]
It is still on time

**Review by DragaDoncila — APPROVED** [review]
Thanks @jules-vanaret this is great 🎉 ! Pulled down the latest changes and played with it a bit and it works as expected for my tracks. I'm also observing the speedup ❤️

**Comment by jni** [comment]
Fantastic work @jules-vanaret, thank you! ❤️‍🔥

**Pull request merged**
